pulsar-client/src/main/java/org/apache/pulsar/client/impl/ControlledClusterFailover.java

```diff
@@ -0,0 +1,277 @@
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@Slf4j
+public class ControlledClusterFailover implements ServiceUrlProvider {
+    private PulsarClient pulsarClient;
+    private volatile String currentPulsarServiceUrl;
+    private volatile ControlledConfiguration currentControlledConfiguration;
+    private final URL pulsarUrlProvider;
+    private final ScheduledExecutorService executor;
+    private long interval;
+    private ObjectMapper objectMapper = null;
+
+    private static final String AUTH_SASL = "org.apache.pulsar.client.impl.auth.AuthenticationSasl";
+    private static final String AUTH_ATHENZ = "org.apache.pulsar.client.impl.auth.AuthenticationAthenz";
+    private static final String AUTH_OAUTH2 = "org.apache.pulsar.client.impl.auth.oauth2.AuthenticationOAuth2";
+    private static final String AUTH_TLS = "org.apache.pulsar.client.impl.auth.AuthenticationTls";
+    private static final String AUTH_TOKEN = "org.apache.pulsar.client.impl.auth.AuthenticationToken";
+    private static final String AUTH_KEY_STORE_TLS = "org.apache.pulsar.client.impl.auth.AuthenticationKeyStoreTls";
+
+    private ControlledClusterFailover(String defaultServiceUrl, String urlProvider, long interval) throws IOException {
+        this.currentPulsarServiceUrl = defaultServiceUrl;
+        this.pulsarUrlProvider = new URL(urlProvider);
+        this.interval = interval;
+        this.executor = Executors.newSingleThreadScheduledExecutor(
+                new DefaultThreadFactory("pulsar-service-provider"));
+    }
+
+    @Override
+    public void initialize(PulsarClient client) {
+        this.pulsarClient = client;
+
+        // start to check service url every 30 seconds
+        this.executor.scheduleAtFixedRate(catchingAndLoggingThrowables(() -> {
+            ControlledConfiguration controlledConfiguration = null;
+            try {
+                controlledConfiguration = fetchControlledConfiguration();
+                if (controlledConfiguration != null
+                        && !Strings.isNullOrEmpty(controlledConfiguration.getServiceUrl())
+                        && !controlledConfiguration.equals(currentControlledConfiguration)) {
+                    log.info("Switch Pulsar service url from {} to {}",
+                            currentControlledConfiguration, controlledConfiguration.toString());
+
+                    Authentication authentication = null;
+                    if (!Strings.isNullOrEmpty(controlledConfiguration.authPluginClassName)) {
+                        String authPluginClassName = controlledConfiguration.authPluginClassName;
+                        Map<String, String> authParams = controlledConfiguration.getAuthParams();
+                        String authParamsString = controlledConfiguration.getAuthParamsString();
+                        String token = controlledConfiguration.getToken();
+
+                        switch (authPluginClassName) {
```

Comment:
I remembered that we can configure the `authPlugin` and `authParams` for all auth plugins, why do we need to case the auth plugins? Does the `AuthenticationFactory.create(authPlugin, authParam)` can not handle all things?

Reply:
done